vscode-extension/src/providers/textDocumentContentProvider.ts

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {TextDocumentContentProvider, Event, Uri, ProviderResult} from 'vscode';
+import {execSync} from 'child_process';
+import * as path from 'path';
+
+const RELAY_CONFIG_SCHEMA_PATH = 'relay-config-schema';
+const PACKAGE_JSON_RELAY_CONFIG_SCHEMA_PATH =
+  'package-json-relay-config-schema';
+
+export class RelayTextDocumentContentProvider
+  implements TextDocumentContentProvider {
+  constructor(private context: any) {}
+
+  static readonly scheme: string = 'relay';
+
+  cachedJsonSchema: string | undefined;
+
+  onDidChange?: Event<Uri> | undefined;
+
+  provideTextDocumentContent(uri: Uri): ProviderResult<string> {
+    if (uri.authority === PACKAGE_JSON_RELAY_CONFIG_SCHEMA_PATH) {
+      return `{"properties": { "relay": { "$ref": "${RelayTextDocumentContentProvider.scheme}://${RELAY_CONFIG_SCHEMA_PATH}", "description": "Relay.js configuration" }}}`;
+    }
+
+    if (uri.authority === RELAY_CONFIG_SCHEMA_PATH) {
+      this.cachedJsonSchema ||= this.loadConfigJsonSchema();
```

Comment:
If the user changes their relay binary path in the config file, does this get invalidated correctly?

Reply:
No, but you have to restart the extension anyways because the binary information is only loaded once at startup atm.